sleap/skeleton.py

```diff
@@ -999,34 +999,153 @@ def to_json(self, node_to_idx: Optional[Dict[Node, int]] = None) -> str:
-        graph = json_graph.node_link_data(indexed_node_graph)
-
-        # SLEAP v1.3.0 added `description` and `preview_image` to `Skeleton`, but saving
-        # these fields breaks data format compatibility. Currently, these are only
-        # added in our custom template skeletons. To ensure backwards data format
-        # compatibilty of user data, we only save these fields if they are not None.
+        # Create global list of nodes with all nodes from all skeletons.
+        nodes_dicts = []
+        node_to_id = {}
+        for node in self.nodes:
+            if node not in node_to_id:
+                print(f'node: {node}')
+                # Note: This ID is not the same as the node index in the skeleton in
+                # legacy SLEAP, but we do not retain this information in the labels, so
+                # IDs will be different.
+                #
+                # The weight is also kept fixed here, but technically this is not
+                # modified or used in legacy SLEAP either.
+                #
+                # TODO: Store legacy metadata in labels to get byte-level compatibility?
+                node_to_id[node] = len(node_to_id)
+                print(f'node_to_id: {node_to_id}')
+                nodes_dicts.append({"name": node.name, "weight": 1.0})
+        print(f'nodes_dicts: {nodes_dicts}')
+
+        # Build links dicts for normal edges.
+        edges_dicts = []
+        for edge_ind, edge in enumerate(self.edges):
+            print(f'edge_ind: {edge_ind}')
+            print(f'edge: {edge}')
+            if edge_ind == 0:
+                edge_type = {
+                    "py/reduce": [
+                        {"py/type": "sleap.skeleton.EdgeType"},
+                        {"py/tuple": [1]},  # 1 = real edge, 2 = symmetry edge
+                    ]
+                }
+                print(f'edge_type: {edge_type}')
+            else:
+                edge_type = {"py/id": 1}
+                print(f'edge_type: {edge_type}')
+
+            # Edges are stored as a list of tuples of nodes
+            # The source and target are the nodes in the tuple (edge) are the first and 
+            # second nodes respectively
+            source = edge[0]
+            print(f'source: {source}')
+            print(f'node_to_id[source]: {node_to_id[source]}')
+            target = edge[1]
+            print(f'target: {target}')
+            print(f'node_to_id[target]: {node_to_id[target]}')
+            edges_dicts.append(
+                {
+                    # Note: Insert idx is not the same as the edge index in the skeleton
+                    # in legacy SLEAP.
+                    "edge_insert_idx": edge_ind,
+                    "key": 0,  # Always 0.
+                    "source": {"py/object": "sleap.skeleton.Node", "py/state": {"name": source.name, "weight": 1.0}},
+                    "target": {"py/object": "sleap.skeleton.Node", "py/state": {"name": target.name, "weight": 1.0}},
```

Comment:
If we are moving the `"nodes"` list above the `"links"` dictionary, then we should be able to just reference the `Node`'s `"py/id"` here (i.e. their index in that `"nodes"` list which is also the `Node`'s index in `Skeleton.nodes`)
```suggestion
                    "source": {"py/id": self.nodes.index(node)},
                    "target": {"py/id": self.nodes.index(node)},
```